tools/read-conf.py

```diff
@@ -7,57 +7,105 @@
+    return conf.get(key)
+
+def get_string_array(conf, key):
+    string = get_string(conf, key)
+    return None if string is None else re.split(r"[, ]+", string)
+
+def get_bool(conf, key):
+    return conf.getboolean(key)
+
+def get_int(conf, key):
+    return conf.getint(key)
+
+def get_int_array(conf, key):
+    array = get_string_array(conf, key)
+    return None if array is None else [int(x) for x in array]
+
+
+dtypes = {
+    # section APY
+    "HTML_URL": get_string,
+    "APY_URL": get_string,
+
+    "SUBTITLE": get_string,
+    "SUBTITLE_COLOR": get_string,
+
+    "ALLOWED_LANGS": get_string_array,
+    "ALLOWED_VARIANTS": get_string_array,
+    "ALLOWED_PAIRS": get_string_array,
+
+    "ENABLED_MODES": get_string_array,
+    "DEFAULT_MODE": get_string,
+
+    "TRANSLATION_CHAINING": get_bool,
+
+    "DEFAULT_LOCALE": get_string,
+
+    "SHOW_NAVBAR": get_string,
+
+    "PIWIK_SITEID": get_string,
+    "PIWIK_URL": get_string,
+
+    "LIST_REQUEST_CACHE_EXPIRY": get_int,
+    "LANGUAGE_NAME_CACHE_EXPIRY": get_int,
+    "LOCALIZATION_CACHE_EXPIRY": get_int,
+    "AVAILABLE_LOCALES_CACHE_EXPIRY": get_int,
+
+    # section PERSISTENCE
+    "DEFAULT_EXPIRY_HOURS": get_int,
+
+    # section TRANSLATOR
+    "UPLOAD_FILE_SIZE_LIMIT": get_int,
+
+    "TRANSLATION_LIST_BUTTONS": get_int,
+    "TRANSLATION_LIST_WIDTH": get_int,
+    "TRANSLATION_LIST_ROWS": get_int,
+    "TRANSLATION_LIST_COLUMNS": get_int,
+    "TRANSLATION_LISTS_BUFFER": get_int,
+
+    "INSTANT_TRANSLATION_URL_DELAY": get_int,
+    "INSTANT_TRANSLATION_PUNCTUATION_DELAY": get_int,
+    "INSTANT_TRANSLATION_DELAY": get_int,
+
+    # section UTIL
+    "TEXTAREA_AUTO_RESIZE_MINIMUM_WIDTH": get_int,
+    "BACK_TO_TOP_BUTTON_ACTIVATION_HEIGHT": get_int,
+    "APY_REQUEST_URL_THRESHOLD_LENGTH": get_int,
+    "DEFAULT_DEBOUNCE_DELAY": get_int,
+
+    "INSTALLATION_NOTIFICATION_REQUESTS_BUFFER_LENGTH": get_int,
+    "INSTALLATION_NOTIFICATION_INDIVIDUAL_DURATION_THRESHOLD": get_int,
+    "INSTALLATION_NOTIFICATION_CUMULATIVE_DURATION_THRESHOLD": get_int,
+    "INSTALLATION_NOTIFICATION_DURATION": get_int,
+}
+
 
 def check_config(conf, result):
     # Some error checking:
     for section in conf.sections():
-        if section not in ['APY', 'REPLACEMENTS']:
+        if section not in ['APY', 'PERSISTENCE', 'REPLACEMENTS', 'TRANSLATOR', 'UTIL']:
             raise configparser.Error("\nUnknown section [%s]" % (section,))
 
+    # TODO: either remove or check for all sections
     apy_diff = set(k.lower() for k in conf['APY'].keys()) - set(k.lower() for k in result.keys())
     if apy_diff:
         raise configparser.Error("\nUnknown key(s) in section [APY]: %s" % (apy_diff,))
 
     return True
 
-def load_conf(filename):
-    conf = configparser.ConfigParser()
-    with open(filename, 'r') as f:
-        conf.read_file(f)
-    conf_APY = conf['APY']
-    result = {
-        'HTML_URL'                       : conf_APY.get('HTML_URL', fallback="http://www.apertium.org"),
```

Comment:
You forgot about the fallbacks this time :)

Reply:
I had them in the custom format file but when I turned it into a config file, I set the initial values to the fallbacks. I will add a `dict` for all non-`None` fallbacks in `read-conf.py`